File: runtime/core-rust/src/pricing/model_resolver.rs
//! Resolve `ModelPricingEntry` with exact match and fallbacks.

use super::types::ModelPricingEntry;

pub fn resolve_entry<'a>(
    entries: &'a [ModelPricingEntry],
    provider: &str,
    model_id: &str,
    warnings: &mut Vec<String>,
) -> Option<&'a ModelPricingEntry> {
    let exact = entries.iter().find(|e| {
        e.provider.eq_ignore_ascii_case(provider) && e.model_id.eq_ignore_ascii_case(model_id)
    });
    if let Some(e) = exact {
        return Some(e);
    }

    if let Some(stripped) = strip_version_suffix_owned(model_id) {
        let v = entries.iter().find(|e| {
            e.provider.eq_ignore_ascii_case(provider)
                && e.model_id.eq_ignore_ascii_case(&stripped)
        });
        if let Some(e) = v {
            warnings.push(format!(
                "pricing: using stripped model id fallback '{stripped}' for '{model_id}'"
            ));
            return Some(e);
        }
    }

    let prefix = entries.iter().find(|e| {
        e.provider.eq_ignore_ascii_case(provider)
            && model_id.starts_with(&e.model_id)
            && e.model_id.len() >= 4
    });
    if let Some(e) = prefix {
        warnings.push(format!(
            "pricing: prefix fallback matched entry '{}' for model '{}'",
            e.model_id, model_id
        ));
        return Some(e);
    }

    let provider_default = entries.iter().find(|e| {
        e.provider.eq_ignore_ascii_case(provider)
            && (e.model_id.eq_ignore_ascii_case("*") || e.model_id.eq_ignore_ascii_case("default"))
    });
    if let Some(e) = provider_default {
        warnings.push(format!(
            "pricing: provider default entry used for model '{model_id}'"
        ));
        return Some(e);
    }

    None
}

fn strip_version_suffix_owned(model_id: &str) -> Option<String> {
    let parts: Vec<&str> = model_id.split('-').collect();
    if parts.len() >= 2 {
        let last = parts[parts.len() - 1];
        if last.chars().all(|c| c.is_ascii_digit() || c == '.' || c == '_') && !last.is_empty() {
            return Some(parts[..parts.len() - 1].join("-"));
        }
    }
    None
}

File: packages/sdk/src/pricing/modelResolver.ts
import type { ModelPricingEntry } from "./types.js";

/**
 * Resolve a pricing entry for provider + model (exact, stripped suffix, prefix, default).
 */
export function resolveModelPricingEntry(
  entries: readonly ModelPricingEntry[],
  provider: string,
  modelId: string,
  warnings: string[],
): ModelPricingEntry | null {
  const p = provider.toLowerCase();
  const m = modelId.trim();
  const exact = entries.find(
    e => e.provider === p && e.modelId.toLowerCase() === m.toLowerCase(),
  );
  if (exact) return exact;

  const stripped = stripVersionSuffix(m);
  if (stripped !== m) {
    const v = entries.find(e => e.provider === p && e.modelId.toLowerCase() === stripped.toLowerCase());
    if (v) {
      warnings.push(`pricing: stripped model id fallback '${stripped}' for '${m}'`);
      return v;
    }
  }

  const prefix = entries.find(
    e => e.provider === p && m.startsWith(e.modelId) && e.modelId.length >= 4,
  );
  if (prefix) {
    warnings.push(`pricing: prefix fallback matched entry '${prefix.modelId}' for model '${m}'`);
    return prefix;
  }

  const def = entries.find(e => e.provider === p && (e.modelId === "*" || e.modelId === "default"));
  if (def) {
    warnings.push(`pricing: provider default entry used for model '${m}'`);
    return def;
  }
  return null;
}

function stripVersionSuffix(modelId: string): string {
  const parts = modelId.split("-");
  if (parts.length < 2) return modelId;
  const last = parts[parts.length - 1] ?? "";
  if (/^[\d._]+$/.test(last) && last.length > 0) {
    return parts.slice(0, -1).join("-");
  }
  return modelId;
}
